README.md
# k8s:fwd — config-based kubernetes multi-port forwarding

A tool for handling port-forwards to multiple services and across multiple clusters.

For Linux (GNU and musl) you can download pre-built binaries from the [Releases](https://github.com/sunsided/k8sfwd/releases)
page. For other platforms the setup is currently based on [cargo] until platform-specific binaries can be provided.
To manually build and install the latest version for your system (or update to it), run:

```shell
cargo install k8sfwd
```

Please note that the application internally relies on `kubectl`, so it needs to be present in your path.
If `kubectl` is not on your path, you may specify it via the `--kubectl` argument or
the `KUBECTL_PATH` environment variable.

Depending on your configuration, you'll be greeted with something along the lines of:

```
██╗░░██╗░█████╗░░██████╗░░░░░███████╗██╗░░░░░░░██╗██████╗
██║░██╔╝██╔══██╗██╔════╝░██╗░██╔════╝██║░░██╗░░██║██╔══██╗
█████═╝░╚█████╔╝╚█████╗░░╚═╝░█████╗░░╚██╗████╗██╔╝██║░░██║
██╔═██╗░██╔══██╗░╚═══██╗░██╗░██╔══╝░░░████╔═████║░██║░░██║
██║░╚██╗╚█████╔╝██████╔╝░╚═╝░██║░░░░░░╚██╔╝░╚██╔╝░██████╔╝
╚═╝░░╚═╝░╚════╝░╚═════╝░░░░░░╚═╝░░░░░░░╚═╝░░░╚═╝░░╚═════╝
k8s:fwd 0.3.0 - a Kubernetes multi-cluster port forwarder
Using kubectl version v1.24.12-dispatcher
Using config from 2 locations

Forwarding to the following targets:
#0 Items API (Staging)
   target:  service/foo.test-api
   context: (default)
   cluster: (default)
#1 Items API (Production)
   target:  pod/foo-59b58f5d68-6t6bh.test-api
   context: (default)
   cluster: production

Spawning child processes:
#0: Error from server (NotFound): pods "foo-59b58f5d68-6t6bh" not found
#0: Process exited with exit status: 1 - will retry in 5 sec
#1: Forwarding from 127.0.0.1:5012 -> 80
#1: Forwarding from 127.0.0.1:46737 -> 8080
#1: Forwarding from [::1]:5012 -> 80
#1: Forwarding from [::1]:46737 -> 8080
#0: Error from server (NotFound): pods "foo-59b58f5d68-6t6bh" not found
#0: Process exited with exit status: 1 - will retry in 5 sec
```

## Command-Line Options

### Filters

Targets can be selected through prefix filters specified on the command-line. Only
targets (and target names) starting with the specified prefixes will be forwarded.
In the following example, services starting with `foo` and `bar` will be selected:

```shell
k8sfwd foo bar
```

Filters can operate in combination with tags as well:

```shell
k8sfwd -t test foo bar
```

### Tags

Targets can be labeled with tags. When `k8sfwd` is started with one or more space-separated
`--tags` parameters, targets are filtered down to match the selection. If multiple values
are specified (e.g. `--tags foo bar`), any matching tag results in the target being selected.
If two tags are combined with a plus sign (e.g. `--tags foo+bar`) only targets matching both
tags are selected.

| Target tags             | `--tags` argument      | Selected |
|-------------------------|------------------------|----------|
| (none)                  | (none)                 | ✅ yes    |
| (none)                  | `--tags some`          | ❌ no     |
| `["foo", "bar", "baz"]` | (none)                 | ✅ yes    |
| `["foo", "bar", "baz"]` | `--tags fubar`         | ❌ no     |
| `["foo", "bar", "baz"]` | `--tags foo bar`       | ✅ yes    |
| `["foo", "bar", "baz"]` | `--tags bar`           | ✅ yes    |
| `["foo", "bar", "baz"]` | `--tags foo+baz`       | ✅ yes    |
| `["foo", "bar", "baz"]` | `--tags foo+fubar`     | ❌ no     |
| `["foo", "bar", "baz"]` | `--tags foo+baz fubar` | ✅ yes    |
| `["fubar"]`             | `--tags foo+baz fubar` | ✅ yes    |
   

## Configuration

The configuration is provided as a YAML file. 

- If one or more files are specified on program launch via the `--file` argument(s), their configuration is loaded.
- If no configuration file is specified, `k8sfwd` will recursively look for a `.k8sfwd` file in 
  - the current directory hierarchy, 
  - your home directory and 
  - your configuration directory, in that order.

Non-target configuration (e.g., retry delays) are always loaded from the hierarchy stated above regardless
of whether a `--file` argument is present. However,  all target configuration that is not directly specified
through a file pointed to by the `--file` argument is ignored.

See [`k8sfwd-example.yaml`](k8sfwd-example.yaml) for an example.

```yaml
---
version: 0.2.0
config:
  # Optional: Number of seconds to wait before attempting to re-establish
  # a broken connection.
  retry_delay_sec: 5.0
targets:
  - name: Test API (Staging)    # Optional, for display purposes.
    target: foo                 # The name of the resource to forward to.
    tags:                       # Optional, for use with `--tags <tag1> <tag2>+<tag3>`
      - integration
    type: service               # Can be service, deployment or pod.
    namespace: bar              # The namespace of the resource.
    context: null               # Optional; will default to current context.
    cluster: null               # Optional; will default to current cluster.
    ports:
      - "5012:80"               # Forward resource port 80 to local port 5012.
      - "8080"                  # Forward resource port 8080 to random local port. 
  - name: Test API (Production)
    target: foo-59b58f5d68-6t6bh
    type: pod
    namespace: bar
    cluster: production
    listen_addrs:               # Select the listen addresses; defaults to `localhost`.
      - "127.1.0.1"
    ports:
      - "5012:80"
```

[cargo]: https://crates.io/


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.
This project uses [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

### Internal

- The code around finding `kubectl` was changed in order to better support the use
  of the `gke-gcloud-auth-plugin` utility.

## [0.3.0] - 2023-07-22

### Added

- [#4](https://github.com/sunsided/k8sfwd/pull/4):
  Added support for configuration files in the user's home and config directories.
- [#5](https://github.com/sunsided/k8sfwd/pull/5):
  Source files from the path hierarchy and special directories are now merged.
- [#7](https://github.com/sunsided/k8sfwd/pull/7):
  Multiple config files can now be specified by repeating the `--file` argument.
- [#8](https://github.com/sunsided/k8sfwd/pull/8):
  Added the `--verbose` command-line option for more detailed information on configuration sources.
- [#9](https://github.com/sunsided/k8sfwd/pull/9):
  Added filter command-line arguments that allows to specify a prefix for
  the loaded targets. Only targets matching the prefix will be forwarded.

### Changed

- The path to the provided or detected source file(s) is now kept relative to the
  current working directory only if it is close. If the file is too many layers
  of nesting away, the canonical path is shown instead of a relative one.

## [0.2.0] - 2023-07-17

### Added

- If only the context or the cluster is specified, the other part will be automatically
  looked up from the current configuration. If a single match is found, its value will
  be explicitly specified to `kubectl`. This should help when changing contexts while
  having a port-forwarding session open as intermittent errors will consistently
  produce the same forwarding rule regardless of the currently active context.
- Added support for tags via the `--tags <tag1> <tag2> ...` command-line option. Only
  targets matching any one of the specified tags will be forwarded.

### Fixed

- Only default to current cluster when neither context nor cluster is specified.
  Previously, specifying only one of the values would result in the other
  value being automatically filled from the currently active context, resulting in
  possibly invalid combinations.

## [0.1.0] - 2023-07-16

### Internal

- 🎉 Initial release with support for hierarchical `.k8sfwd` detection.

[0.3.0]: https://github.com/sunsided/k8sfwd/releases/tag/0.3.0
[0.2.0]: https://github.com/sunsided/k8sfwd/releases/tag/0.2.0
[0.1.0]: https://github.com/sunsided/k8sfwd/releases/tag/0.1.0


src/portfwd.rs
use lazy_static::lazy_static;
use semver::Version;
use serde::de::Error;
use serde::{Deserialize, Deserializer};
use std::fs::File;
use std::io;
use std::io::Read;
use std::net::IpAddr;

lazy_static! {
    pub static ref LOWEST_SUPPORTED_VERSION: Version = Version::new(0, 1, 0);
    pub static ref HIGHEST_SUPPORTED_VERSION: Version = Version::new(0, 1, 0);
}

#[derive(Debug, Deserialize)]
pub struct PortForwardConfigs {
    pub version: Version,
    pub targets: Vec<PortForwardConfig>,
}

#[derive(Debug, Deserialize)]
pub struct PortForwardConfig {
    /// An optional name used to refer to this configuration.
    pub name: Option<String>,
    /// The name of the kubeconfig context to use.
    pub context: Option<String>,
    /// The name of the kubeconfig cluster to use.
    pub cluster: Option<String>,
    /// The addresses or host names to listen on; must be an IP address or `localhost`.
    #[serde(default, deserialize_with = "deserialize_listen_addrs")]
    pub listen_addrs: Vec<String>,
    /// The namespace to forward to, e.g. `default`.
    #[serde(default = "default_namespace")]
    pub namespace: String,
    /// The type of resource to forward to.
    #[serde(default)]
    pub r#type: ResourceType,
    /// The name of the resource to forward to.
    pub target: String,
    /// The port to forward.
    pub ports: Vec<Port>,
}

/// The type of resource to forward to.
#[derive(Debug, Deserialize)]
pub enum ResourceType {
    #[serde(rename = "service")]
    Service,
    #[serde(rename = "deployment")]
    Deployment,
    #[serde(rename = "pod")]
    Pod,
}

/// A port to forward.
#[derive(Debug)]
pub struct Port {
    /// The local port to forward to.
    pub local: Option<u16>,
    /// The remote port to forward to.
    pub remote: u16,
}

impl Default for ResourceType {
    fn default() -> Self {
        Self::Service
    }
}

fn default_namespace() -> String {
    "default".to_string()
}

impl ResourceType {
    pub fn to_arg(&self) -> &'static str {
        match self {
            ResourceType::Service => "service",
            ResourceType::Deployment => "deployment",
            ResourceType::Pod => "pod",
        }
    }
}

impl IntoIterator for PortForwardConfigs {
    type Item = PortForwardConfig;
    type IntoIter = std::vec::IntoIter<Self::Item>;

    fn into_iter(self) -> Self::IntoIter {
        self.targets.into_iter()
    }
}

impl PortForwardConfigs {
    pub fn is_supported_version(&self) -> bool {
        #[allow(clippy::absurd_extreme_comparisons)]
        if self.version < *LOWEST_SUPPORTED_VERSION || self.version > *HIGHEST_SUPPORTED_VERSION {
            false
        } else {
            true
        }
    }
}

/// Parses a vector of IP addresses or the literal `localhost`.
fn deserialize_listen_addrs<'de, D>(deserializer: D) -> Result<Vec<String>, D::Error>
where
    D: Deserializer<'de>,
{
    #[derive(Deserialize)]
    struct Wrapper(#[serde(deserialize_with = "deserialize_listen_addr")] String);

    let v = Vec::deserialize(deserializer)?;
    Ok(v.into_iter().map(|Wrapper(a)| a).collect())
}

/// Parses an IPv4 or IPv6 address or the literal `localhost`.
fn deserialize_listen_addr<'de, D>(deserializer: D) -> Result<String, D::Error>
where
    D: Deserializer<'de>,
{
    let buf = String::deserialize(deserializer)?;

    if buf == "localhost" {
        return Ok(buf);
    }

    if buf.starts_with('[') && buf.ends_with(']') {
        let ip = &buf[1..(buf.len() - 1)];
        return if ip.parse::<IpAddr>().is_ok() {
            Ok(buf)
        } else {
            Err(Error::custom(format!(
                "An invalid IPv6 address was specified: {buf}"
            )))
        };
    }

    if buf.parse::<IpAddr>().is_ok() {
        return Ok(buf);
    }

    Err(Error::custom(
        "Listen address must be either \"localhost\" or a valid IP address",
    ))
}

impl<'de> Deserialize<'de> for Port {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct PortVisitor;

        impl<'de> serde::de::Visitor<'de> for PortVisitor {
            type Value = Port;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                formatter.write_str("a string or an object")
            }

            fn visit_i16<E>(self, remote: i16) -> Result<Self::Value, E>
            where
                E: Error,
            {
                if remote <= 0 {
                    return Err(E::custom("Invalid port number: value must be positive"));
                }

                Ok(Port {
                    local: None,
                    remote: remote as _,
                })
            }

            fn visit_u16<E>(self, remote: u16) -> Result<Self::Value, E>
            where
                E: Error,
            {
                if remote == 0 {
                    return Err(E::custom("Invalid port number: value must be positive"));
                }

                Ok(Port {
                    local: None,
                    remote,
                })
            }

            fn visit_u64<E>(self, remote: u64) -> Result<Self::Value, E>
            where
                E: Error,
            {
                if remote == 0 {
                    return Err(E::custom("Invalid port number: value must be positive"));
                }

                if remote > u16::MAX as _ {
                    return Err(E::custom(
                        "Invalid port number: value must be smaller than or equal to 65535",
                    ));
                }

                Ok(Port {
                    local: None,
                    remote: remote as _,
                })
            }

            fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
            where
                E: Error,
            {
                // Split the string by ':' and parse the numbers
                let parts: Vec<&str> = s.split(':').collect();
                match parts[..] {
                    [local, remote] => {
                        let local = match local {
                            "" => None,
                            value => Some(value.parse::<u16>().map_err(E::custom)?),
                        };
                        let remote = remote.parse::<u16>().map_err(E::custom)?;

                        Ok(Port { local, remote })
                    }
                    [remote] => {
                        let remote = remote.parse::<u16>().map_err(E::custom)?;
                        Ok(Port {
                            local: None,
                            remote,
                        })
                    }
                    _ => Err(E::custom("Invalid string format")),
                }
            }

            fn visit_map<M>(self, mut map: M) -> Result<Self::Value, M::Error>
            where
                M: serde::de::MapAccess<'de>,
            {
                // Deserialize the JSON object
                let mut local = None;
                let mut remote = None;

                while let Some(key) = map.next_key::<String>()? {
                    match key.as_str() {
                        "local" => {
                            if local.is_some() {
                                return Err(Error::duplicate_field("local"));
                            }
                            local = Some(map.next_value()?);
                        }
                        "remote" => {
                            if remote.is_some() {
                                return Err(Error::duplicate_field("remote"));
                            }
                            remote = Some(map.next_value()?);
                        }
                        _ => return Err(Error::unknown_field(&key, &["local", "remote"])),
                    }
                }

                Ok(Port {
                    local,
                    remote: remote.ok_or_else(|| Error::missing_field("remote"))?,
                })
            }
        }

        deserializer.deserialize_any(PortVisitor)
    }
}

pub trait FromYaml {
    fn into_configuration(self) -> Result<PortForwardConfigs, FromYamlError>;
}

impl FromYaml for File {
    fn into_configuration(mut self) -> Result<PortForwardConfigs, FromYamlError> {
        let mut contents = String::new();
        self.read_to_string(&mut contents)?;
        Ok(serde_yaml::from_str(&contents)?)
    }
}

#[derive(Debug, thiserror::Error)]
pub enum FromYamlError {
    #[error(transparent)]
    InvalidConfiguration(#[from] serde_yaml::Error),
    #[error(transparent)]
    FileReadFailed(#[from] io::Error),
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_port_from_object() {
        let input = r"
            local: 5012
            remote: 80
        ";

        let port: Port = serde_yaml::from_str(input).unwrap();
        assert_eq!(port.local, Some(5012));
        assert_eq!(port.remote, 80);
    }

    #[test]
    fn test_auto_port_from_object() {
        let input = "remote: 80";
        let port: Port = serde_yaml::from_str(input).unwrap();
        assert_eq!(port.local, None);
        assert_eq!(port.remote, 80);
    }

    #[test]
    fn test_port_from_string() {
        let port: Port = serde_yaml::from_str("5012:80").unwrap();
        assert_eq!(port.local, Some(5012));
        assert_eq!(port.remote, 80);
    }

    #[test]
    fn test_auto_port_from_string() {
        let port: Port = serde_yaml::from_str("\":80\"").unwrap();
        assert_eq!(port.local, None);
        assert_eq!(port.remote, 80);
    }

    #[test]
    fn test_auto_port_from_string_2() {
        let port: Port = serde_yaml::from_str("\"80\"").unwrap();
        assert_eq!(port.local, None);
        assert_eq!(port.remote, 80);
    }

    #[test]
    fn test_auto_port_from_string_3() {
        let port: Port = serde_yaml::from_str(":80").unwrap();
        assert_eq!(port.local, None);
        assert_eq!(port.remote, 80);
    }

    #[test]
    fn test_auto_port_from_string_4() {
        let port: Port = serde_yaml::from_str("80").unwrap();
        assert_eq!(port.local, None);
        assert_eq!(port.remote, 80);
    }

    #[test]
    fn test_listen_ip_and_localhost() {
        serde_yaml::from_str::<PortForwardConfig>(
            r#"
            target: foo
            listen_addrs:
              - "127.0.0.1"
              - "[::1]"
              - "localhost"
            ports:
              - "1234:5678"
        "#,
        )
        .expect("configuration is valid");
    }

    #[test]
    fn test_listen_invalid_host() {
        serde_yaml::from_str::<PortForwardConfig>(
            r#"
            target: foo
            listen_addrs:
              - "foo"
            ports:
              - "1234:5678"
        "#,
        )
        .expect_err("literal host names must be exactly `localhost`");
    }

    #[test]
    fn test_listen_invalid_ipv4() {
        serde_yaml::from_str::<PortForwardConfig>(
            r#"
            target: foo
            listen_addrs:
              - "127.0.0.256"
            ports:
              - "1234:5678"
        "#,
        )
        .expect_err("the IPv6 address is invalid");
    }

    #[test]
    fn test_listen_invalid_ipv6() {
        serde_yaml::from_str::<PortForwardConfig>(
            r#"
            target: foo
            listen_addrs:
              - "[fe80:2030:31:24]"
            ports:
              - "1234:5678"
        "#,
        )
        .expect_err("the IPv6 address is invalid");
    }

    #[test]
    fn test_entire_config() {
        let config = r#"
            version: 0.1.0
            targets:
              - name: Test API (Staging)
                target: foo
                type: service
                namespace: bar
                context: null
                cluster: null
                listen_addrs:
                  - "127.0.0.1"
                ports:
                  - "5012:80"
                  - 8080
              - name: Test API (Production)
                target: foo
                type: service
                namespace: bar
                cluster: production
                listen_addrs:
                  - "127.1.0.1"
                ports:
                  - "5012:80"
        "#;

        let config: PortForwardConfigs = serde_yaml::from_str(config).unwrap();
        assert_eq!(config.targets.len(), 2);
    }
}
